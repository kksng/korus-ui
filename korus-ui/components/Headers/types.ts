import * as React from 'react';

export interface HeaderProps extends React.HTMLAttributes<HTMLHeadingElement> {
  [x: string]: unknown,
  ref?: React.Ref<HeaderRefCurrent>,
  shouldRender?: boolean,
}

export interface HeaderRefCurrent {
  wrapper: HTMLHeadingElement | null,
}
