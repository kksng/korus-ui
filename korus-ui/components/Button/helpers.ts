import { Field, Form, FormsObject } from '../Validation/types';

export const fromFormArraytoFormObject = (forms: Form[]): FormsObject => forms.reduce((
  acc: FormsObject,
  currentForm,
) => {
  const fieldsObject = currentForm.fields.reduce((
    fieldAcc: {[x: string]: Field},
    currentField,
  ): {[fieldName: string]: Field} => {
    fieldAcc[currentField.name] = currentField;
    return fieldAcc;
  }, {});

  acc[currentForm.name] = fieldsObject;
  return acc;
}, {});

// Helper gets DOMRect of Element.
// If Element's height is 0, helper searches for nearest parent with height > 0
export const getElementRect = (element: HTMLElement): DOMRect => {
  const rect = element.getBoundingClientRect();

  if (rect.height === 0 && element.parentElement) {
    return getElementRect(element.parentElement);
  }
  return rect;
};
