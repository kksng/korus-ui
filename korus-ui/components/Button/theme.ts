/* eslint-disable key-spacing */
export const defaultButtonTheme = {
  disabled:                 'disabled',
  loading:                  'loading',
  wrapper:                  'button-wrapper',
};
/* eslint-enable key-spacing */
