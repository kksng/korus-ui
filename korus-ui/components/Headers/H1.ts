import * as React from 'react';
import { htmlTagFactory } from '../../src/HTMLTagsFactory';
import { HeaderProps } from './types';

export const H1 = htmlTagFactory('H1') as React.FC<HeaderProps>;
