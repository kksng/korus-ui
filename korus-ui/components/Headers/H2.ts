import * as React from 'react';
import { htmlTagFactory } from '../../src/HTMLTagsFactory';
import { HeaderProps } from './types';

export const H2 = htmlTagFactory('H2') as React.FC<HeaderProps>;
