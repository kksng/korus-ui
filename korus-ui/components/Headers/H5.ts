import * as React from 'react';
import { htmlTagFactory } from '../../src/HTMLTagsFactory';
import { HeaderProps } from './types';

export const H5 = htmlTagFactory('H5') as React.FC<HeaderProps>;
