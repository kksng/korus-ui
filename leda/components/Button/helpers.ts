import { Field, Form, FormsObject } from '../Validation/types';

export const fromFormArraytoFormObject = (forms: Form[]): FormsObject => forms.reduce((
  acc: FormsObject,
  currentForm,
) => {
  const fieldsObject = currentForm.fields.reduce((
    fieldAcc: {[x: string]: Field},
    currentField,
  ): {[fieldName: string]: Field} => {
    fieldAcc[currentField.name] = currentField;
    return fieldAcc;
  }, {});

  acc[currentForm.name] = fieldsObject;
  return acc;
}, {});
