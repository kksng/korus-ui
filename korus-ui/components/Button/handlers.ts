import * as React from 'react';
import { isFunction, intersection } from 'lodash';
import { getForms, validate } from '../Validation';
import { ButtonProps } from './types';
import { fromFormArraytoFormObject, getElementRect } from './helpers';
import { form as LedaForm } from '../../form';

export const createClickHandler = (props: ButtonProps) => (ev: React.MouseEvent<HTMLButtonElement>): void => {
  const {
    onClick, onValidationFail, isDisabled, isLoading, scrollDelay,
    scrollOffset, form: formProp, shouldScrollToInvalidFields, type,
  } = props;

  if (isDisabled || isLoading) return; // если кнопка отключена или в состояниии загрузки, прервать выполнение функции

  // если кнопка участвует в валидации форм
  if (formProp) {
    const buttonFormNames = Array.isArray(formProp) ? formProp : [formProp];
    const formNames = getForms().map((form) => form.name);
    const validButtonFormNames = intersection(formNames, buttonFormNames);

    if (type === 'reset') {
      validButtonFormNames.map((formName) => LedaForm(formName).reset());
      return;
    }

    const isEachFormValid = validButtonFormNames
      .map((currentForm) => validate(currentForm)) // validate all forms
      .every((item) => item); // tell me if all of them are valid

    if (!isEachFormValid) {
      const invalidForms = getForms()
        .filter((currentForm) => validButtonFormNames.includes(currentForm.name))
        .filter((currentForm) => currentForm.fields.some((field) => !field.isValid));

      if (isFunction(onValidationFail)) {
        onValidationFail({ ...ev, invalidForms });
      }

      if (shouldScrollToInvalidFields && invalidForms.length > 0) {
        const firstInvalidForm = invalidForms[0];
        const firstInvalidField = firstInvalidForm.fields.find((field) => field.isValid === false);

        // get all form elements
        const formElementsArray = [
          ...document.querySelectorAll(`[form=${firstInvalidForm.name}], [data-form=${firstInvalidForm.name}]`),
        ] as HTMLElement[];

        // find first invalid element by name attribute (unique for form elements)
        const invalidElement = formElementsArray.find((fieldElement) => (
          fieldElement.getAttribute('name') === firstInvalidField?.name
        )) as HTMLElement;

        // wait for data update
        setTimeout(() => {
          if (invalidElement) {
            const invalidElementRect = getElementRect(invalidElement);
            const isIE = !!(document as any).documentMode || /Edge/.test(navigator.userAgent);
            const offset = invalidElementRect.top - (scrollOffset ?? 0);
            if (isIE) {
              window.scrollBy(0, offset);
            } else {
              window.scrollBy({
                behavior: 'smooth',
                top: offset,
              });
            }
          }
        }, scrollDelay ?? 0);
      }

      return;
    }
  }

  if (isFunction(onClick)) {
    if (!formProp) {
      onClick(ev);
      return;
    }

    const forms = getForms(formProp);
    const formsObject = fromFormArraytoFormObject(forms);

    const customEvent = {
      ...ev,
      form: formsObject,
      forms,
    };

    onClick(customEvent);
  }
};
