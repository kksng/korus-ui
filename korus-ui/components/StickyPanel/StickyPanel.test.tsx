import React from 'react';
import { render } from '@testing-library/react';
import { StickyPanel } from './index';
import { Button } from '../Button';
import { H1 } from '../Headers';

describe('StickyPanel SNAPSHOTS', () => {
  beforeEach(() => {
    (window as any).MutationObserver = function () {
      this.observe = jest.fn();
      this.disconnect = jest.fn();
    };
  });

  test('render', () => {
    const wrapper = render((
      <StickyPanel>
        <Button _success>OK</Button>
        <Button>Cancel</Button>
      </StickyPanel>
    ));
    expect(document.querySelector('div.stickypanel-wrapper')).not.toBeNull();
    expect(document.querySelector('div.stickypanel-container')).not.toBeNull();
    expect(wrapper.container).toMatchSnapshot();
  });
});

describe('StickyPanel ATTRIBUTES', () => {
  test('children', () => {
    const wrapper = render((
      <StickyPanel>
        <H1>Would you like to continue?</H1>
        <Button _success>OK</Button>
        <Button>Cancel</Button>
      </StickyPanel>
    ));
    expect(wrapper.container).toHaveTextContent('Would you like to continue?');
    expect(wrapper.container).toHaveTextContent('OK');
    expect(wrapper.container).toHaveTextContent('Cancel');
  });
});
