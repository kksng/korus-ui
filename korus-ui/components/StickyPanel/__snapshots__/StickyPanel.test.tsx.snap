// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StickyPanel SNAPSHOTS render 1`] = `
<div>
  <div
    class="stickypanel-wrapper"
    style="height: 0px;"
  >
    <div
      class="stickypanel-container"
    >
      <button
        class="button-wrapper success"
        type="button"
      >
        OK
      </button>
      <button
        class="button-wrapper"
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
